File: pt7/1302.deepest-leaves-sum.py
#
# @lc app=leetcode id=1302 lang=python3
#
# [1302] Deepest Leaves Sum
#
# https://leetcode.com/problems/deepest-leaves-sum/description/
#
# algorithms
# Medium (86.53%)
# Likes:    4847
# Dislikes: 127
# Total Accepted:    407.4K
# Total Submissions: 470.8K
# Testcase Example:  '[1,2,3,4,5,null,6,7,null,null,null,null,8]'
#
# Given the root of a binary tree, return the sum of values of its deepest
# leaves.
# 
# Example 1:
# 
# 
# Input: root = [1,2,3,4,5,null,6,7,null,null,null,null,8]
# Output: 15
# 
# 
# Example 2:
# 
# 
# Input: root = [6,7,8,2,7,1,3,9,null,1,4,null,null,null,5]
# Output: 19
# 
# 
# 
# Constraints:
# 
# 
# The number of nodes in the tree is in the range [1, 10^4].
# 1 <= Node.val <= 100
# 
# 
#

# @lc code=start
from __future__ import annotations

from collections import deque
from typing import Optional


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        queue = deque([root])
        level_sum = 0

        while queue:
            level_sum = 0
            for _ in range(len(queue)):
                node = queue.popleft()
                level_sum += node.val

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

        return level_sum
# @lc code=end

#
# Interview explanation
# ---------------------
# Idea:
# We need the sum of the deepest level only. A level-order traversal naturally
# visits the tree one depth at a time, so we keep replacing `level_sum` with the
# sum of the current layer. When the BFS finishes, the last computed layer is
# exactly the deepest leaves.
#
# Data structure:
# `deque` is the right queue for BFS because popping from the left is O(1).
# A Python list would make `pop(0)` O(n), which is unnecessary overhead.
#
# Walkthrough:
# 1. Put the root in the queue.
# 2. For each BFS layer, reset `level_sum` to 0.
# 3. Consume exactly the nodes that were in the queue at the start of that
#    layer, add their values, and enqueue their children.
# 4. Return the final `level_sum`.
#
# Edge cases:
# - Empty tree: return 0 defensively, even though LeetCode gives a non-empty
#   tree for this problem.
# - Single node: one BFS layer, so the answer is the root value.
# - Skewed tree: still works because each layer contains one node.
#
# Complexity:
# - Time: O(n), every node is visited once.
# - Space: O(w), where w is the maximum width of the tree held in the queue.
#
# Tests to discuss:
# - [1,2,3,4,5,null,6,7,null,null,null,null,8] -> 15.
# - [6,7,8,2,7,1,3,9,null,1,4,null,null,null,5] -> 19.
# - [1] -> 1.

File: pt7/1302.deepest-leaves-sum.java
import java.util.ArrayDeque;
import java.util.Queue;

/*
 * LeetCode 1302 - Deepest Leaves Sum
 *
 * Definition for a binary tree node is provided by LeetCode:
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int deepestLeavesSum(TreeNode root) {
        if (root == null) {
            return 0;
        }

        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);
        int levelSum = 0;

        while (!queue.isEmpty()) {
            levelSum = 0;
            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                levelSum += node.val;

                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
        }

        return levelSum;
    }
}

/*
 * Interview explanation
 * ---------------------
 * Algorithm:
 * We need only the last level of the tree. A breadth-first search naturally
 * visits nodes level by level, so we overwrite `levelSum` for every level. When
 * the queue becomes empty, the most recent sum is the deepest leaves sum.
 *
 * Java data structures:
 * `Queue<TreeNode>` backed by `ArrayDeque` gives O(1) enqueue and dequeue.
 * `LinkedList` would also work, but `ArrayDeque` is the usual Java choice for
 * queue/stack behavior when we do not need random access.
 *
 * Edge cases:
 * - Empty root returns 0 defensively.
 * - Single-node tree returns the root value.
 * - Skewed tree works because each BFS layer simply has one node.
 *
 * Complexity:
 * Time O(n), every node is visited once.
 * Space O(w), where w is the maximum tree width stored in the queue.
 */
